app/database/build_store.py

```diff
@@ -0,0 +1,307 @@
+            error_message=build._error_message,
+            postbuild_tasks_are_finished=bool(build._postbuild_tasks_are_finished),
+            setup_failures=build.setup_failures,
+            timing_file_path=build._timing_file_path,
+            build_artifact_dir=build_artifact_dir,
+            build_parameters=json.dumps(build._build_request.build_parameters()),
+            state=build._status(),
+            queued_ts=fsm_timestamps['queued'],
+            finished_ts=fsm_timestamps['finished'],
+            prepared_ts=fsm_timestamps['prepared'],
+            preparing_ts=fsm_timestamps['preparing'],
+            error_ts=fsm_timestamps['error'],
+            canceled_ts=fsm_timestamps['canceled'],
+            building_ts=fsm_timestamps['building']
+        )
+        session.add(build_schema)
+
+        # Commit this first to get the build_id created by the database
+        # We use this build_id to store the other parts of a Build object
+        session.commit()
+        build_id = build_schema.build_id
+
+        # FailedArtifactDirectories
+        if build._build_artifact is not None:
+            for directory in build._build_artifact._get_failed_artifact_directories():
+                failed_artifact_directory = FailedArtifactDirectoriesSchema(
+                    build_id=build_id,
+                    failed_artifact_directory=directory
+                )
+                session.add(failed_artifact_directory)
+
+        # FailedSubjobAtomPairs
+        if build._build_artifact is not None:
+            for subjob_id, atom_id in build._build_artifact.get_failed_subjob_and_atom_ids():
+                failed_subjob_and_atom_ids = FailedSubjobAtomPairsSchema(
+                    build_id=build_id,
+                    subjob_id=subjob_id,
+                    atom_id=atom_id
+                )
+                session.add(failed_subjob_and_atom_ids)
+
+        # Subjobs
+        subjobs = build._all_subjobs_by_id
+        for subjob_id in subjobs:
+            subjob = build._all_subjobs_by_id[subjob_id]
+            subjob_schema = SubjobsSchema(
+                subjob_id=subjob_id,
+                build_id=build_id,
+                completed=subjob.completed
+            )
+            session.add(subjob_schema)
+
+            # Atoms
+            for atom in subjob._atoms:
+                atom_schema = AtomsSchema(
+                    atom_id=atom.id,
+                    build_id=build_id,
+                    subjob_id=subjob_id,
+                    command_string=atom.command_string,
+                    expected_time=atom.expected_time,
+                    actual_time=atom.actual_time,
+                    exit_code=atom.exit_code,
+                    state=atom.state,
+                )
+                session.add(atom_schema)
+
+        session.commit()
+        return build_id
+
+    @classmethod
+    def _reconstruct_build(cls, build_id, allow_incompleted_builds=False) -> Build:
```

Comment:
Oh good catch, I don't think we'd ever use this method anywhere -- we could remove it